rfcs/20221209-data-cleaning.md

```diff
@@ -0,0 +1,26 @@
+# RFC: Cleaning up the upstream database
+
+One of the steps of the [data refresh process for images][data-refresh] is cleaning data that is not fit for production. This process runs weekly as Airflow DAGs, the cleaned data is only saved to the API database (DB), which is replaced when the data refresh finishes, so it needs to be cleaned every time. This cleaning step takes up to 16 hours lately. It has been stable in duration given the strategy changed to perform the validations and transformations of the data upfront, when pulling from providers rather than when copying the upstream catalog DB into the API DB, but images ingested prior to this change remained untouched.
+
+The reason why the cleaning step is only applied to images and not to audio records is related to this change. The audio media type was integrated into Openverse after the creation of the MediaStore classes, which are the ones performing general and media-specific validations of the data pulled from providers.
+
+As seen, this is a costly process in terms of time and CPU resources, and it is inefficient to repeat it every time. Here I want to propose solving it in origin by doing a one-time pass by the image table of the upstream DB. We can follow the approach proposed by a previous Data Engineer contributor to [clean preexisting data using ImageStore][imagestore], see the details explained in the pull request and linked issue. The advantage of this way is that we apply not only the [cleaning steps defined in the ingestion server][cleanin-steps-is] but all the extra validations which were added later. I can foresee us using the same workflow in future opportunities if we require applying new modifications to the existing data, as long as its feasibility is confirmed.
+
+## Plan
+
+After evaluating what needs to be done to get this rid of the duplicated cleaning step, I collected the relevant issues of the openverse-catalog into a new milestone, [Data cleaning unification][milestone], and determined a rough plan to solve them:
+
+    1. Include the ingestion server cleaning steps in the ImageStore class
+        - correct tags
+        - correct URI protocol
+    2. Create and run an image_cleaner workflow as described above
+    3. Remove the cleanup step for images from the ingestion process
```

Comment:
Thanks for pulling the issues into the milestone and for writing the steps to the plan. Can you confirm whether the milestone's issues are ordered according to this plan? It's not clear to me from the issue titles which of them correspond to the relevant parts of the plan. Some of them seem easy enough to draw to a specific step of this plan, others less so,

Would it make sense to link the issues to the step described here to make that clearer? Additionally, do the issues in the milestone need to be re-written or re-phrased to match the intention of this RFC? Forgive me if I'm missing context that pulls them together, I may just be misunderstanding the language used in the issues. This one for example: https://github.com/WordPress/openverse-catalog/issues/510. Does "when loading ingested data" mean adding it to the ingestion server project as part of the data refresh? Or would it be changed to reference adding it to the ImageStore cleaning routine? If they need re-writing I guess that would happen after the RFC is approved?

Again, sorry if I'm misunderstanding something here and please let me know if I am.

Reply:
The issues have not been rewritten to reflect the intended changes. In the case of that issue you mentioned, it seems that was the goal at the moment of creation but adding more steps to the ingestion server is precisely what we would want to avoid here. So yes, I can re-phrase them for sure, if the plan makes sense for the team.

Also, I'm noticing it needs one especially for recovering the `image_cleaner` DAG, a central piece! Aside from that, I don't think it requires a strict order aside from the one described here in the RFC. Issues related to validations need to get resolved before running the cleaner DAG, and removing the cleanup steps from the ingestion server should be the last one.